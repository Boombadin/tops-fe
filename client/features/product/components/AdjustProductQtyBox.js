import { IncrementInputBox } from '@central-tech/core-ui';
import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';

const IncrementInputBoxStyled = styled(IncrementInputBox)`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;

  box-sizing: content-box;
  height: ${({ type }) => (type == 'mini' ? '20px' : '43px')};
  background-color: #fff;
  border: 3px solid #f60403;
  border-radius: 6px;
  color: #fff;
  line-height: ${({ type }) => (type == 'mini' ? '20px' : '43px')};

  input {
    color: #f60403;
    font-size: ${({ type }) => (type == 'mini' ? '12px' : '15px')};
    font-weight: 700;
    border: none;

    width: ${({ type }) => type == 'mini' && '100%'};
    height: ${({ type }) => type == 'mini' && 'inherit'};
  }
`;

const ButtonStyled = styled.button`
  width: ${({ type }) => (type == 'mini' ? '18.5px' : '35px')};
  height: ${({ type }) => (type == 'mini' ? '20px' : '43px')};

  font-size: 16px;
  background-color: #f60403;
  border: none;
  color: #fff;
  cursor: pointer;
  text-align: center;
`;

function AdjustProductQtyBox({
  defaultQty,
  stockQty,
  maxSaleQty,
  gtm,
  onClickIncreaseQtyBtn,
  onClickDecreaseQtyBtn,
  onClickIncreaseQtyDisabledBtn,
  onChangeProductQtyCompleted,
  type,
}) {
  return (
    <IncrementInputBoxStyled
      data-testid="adjust-product-qty-box"
      defaultValue={defaultQty}
      editable={false}
      minValue={0}
      maxValue={stockQty < maxSaleQty ? stockQty : maxSaleQty}
      onUpdateCompleted={onChangeProductQtyCompleted}
      type={type}
      renderButton={direction =>
        direction === 'inc' ? (
          <ButtonStyled
            data-testid="adjust-product-qty-box-increase-btn"
            onClick={onClickIncreaseQtyBtn}
            type={type}
            {...gtm}
          >
            +
          </ButtonStyled>
        ) : (
          <ButtonStyled
            data-testid="adjust-product-qty-box-decrease-btn"
            onClick={onClickDecreaseQtyBtn}
            type={type}
            {...gtm}
          >
            -
          </ButtonStyled>
        )
      }
      renderButtonDisabled={direction =>
        direction === 'inc' ? (
          <ButtonStyled
            data-testid="adjust-product-qty-box-increase-btn"
            type={type}
            onClick={onClickIncreaseQtyDisabledBtn}
            {...gtm}
          >
            +
          </ButtonStyled>
        ) : (
          <ButtonStyled
            data-testid="adjust-product-qty-box-decrease-btn"
            disabled
            type={type}
            {...gtm}
          >
            -
          </ButtonStyled>
        )
      }
    />
  );
}

AdjustProductQtyBox.propTypes = {
  defaultQty: PropTypes.number,
  stockQty: PropTypes.number,
  maxSaleQty: PropTypes.number,
  gtm: PropTypes.object,
  onClickIncreaseQtyBtn: PropTypes.func.isRequired,
  onClickDecreaseQtyBtn: PropTypes.func.isRequired,
  onChangeProductQtyCompleted: PropTypes.func.isRequired,
  type: PropTypes.oneOf(['default', 'mini']).isRequired,
};

AdjustProductQtyBox.defaultProps = {
  defaultQty: 1,
  stockQty: null,
  maxSaleQty: null,
  gtm: {},
  type: 'default',
};

export default React.memo(AdjustProductQtyBox);
