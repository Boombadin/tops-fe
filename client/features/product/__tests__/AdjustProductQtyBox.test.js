import { cleanup, fireEvent, render } from '@testing-library/react';
import React from 'react';

import AdjustProductQtyBox from '@client/features/product/components/AdjustProductQtyBox';

describe('features/product/components/AdjustProductQtyBox', () => {
  let props = {};
  beforeEach(() => {
    props = {
      defaultQty: 1,
      type: 'default',
      stockQty: 100,
      maxSaleQty: 100,
      onChangeProductQtyCompleted: jest.fn(),
      onClickIncreaseQtyBtn: jest.fn(),
      onClickDecreaseQtyBtn: jest.fn(),
    };
  });
  afterEach(cleanup);

  test('it should render default qty label from prop', async () => {
    props.defaultQty = 10;
    const component = render(<AdjustProductQtyBox {...props} />);
    component.getByDisplayValue(props.defaultQty.toString());

    expect(component).toMatchSnapshot();
  });
  test('it should render readonly input', () => {
    const component = render(<AdjustProductQtyBox {...props} />);
    expect(
      component.getByDisplayValue(props.defaultQty.toString()),
    ).toHaveAttribute('readonly');

    expect(component).toMatchSnapshot();
  });
  test('it should display increased value on click increase btn', () => {
    const component = render(<AdjustProductQtyBox {...props} />);
    fireEvent.click(
      component.getByTestId('adjust-product-qty-box-increase-btn'),
    );
    component.getByDisplayValue((props.defaultQty + 1).toString());
  });
  test('it should display decreased value on click decrease btn', () => {
    const component = render(<AdjustProductQtyBox {...props} />);
    fireEvent.click(
      component.getByTestId('adjust-product-qty-box-decrease-btn'),
    );
    component.getByDisplayValue((props.defaultQty - 1).toString());
  });
  test('it should not decrease display value to less than 0', () => {
    props.defaultQty = 0;
    const component = render(<AdjustProductQtyBox {...props} />);
    fireEvent.click(
      component.getByTestId('adjust-product-qty-box-decrease-btn'),
    );
    component.getByDisplayValue(props.defaultQty.toString());
  });
  test('it should not change the display value on click increase btn when currentQty equal to stockQty', () => {
    props.stockQty = 2;
    const component = render(<AdjustProductQtyBox {...props} />);
    expect(
      component.getByTestId('adjust-product-qty-box-increase-btn'),
    ).not.toHaveAttribute('disabled');

    fireEvent.click(
      component.getByTestId('adjust-product-qty-box-increase-btn'),
    );
    component.getByDisplayValue((props.defaultQty + 1).toString());

    fireEvent.click(
      component.getByTestId('adjust-product-qty-box-increase-btn'),
    );
    component.getByDisplayValue((props.defaultQty + 1).toString());
  });
  test('it should not change the display value on click increase btn when currentQty equal to maxSaleQty', () => {
    props.maxSaleQty = 2;
    const component = render(<AdjustProductQtyBox {...props} />);
    expect(
      component.getByTestId('adjust-product-qty-box-increase-btn'),
    ).not.toHaveAttribute('disabled');

    fireEvent.click(
      component.getByTestId('adjust-product-qty-box-increase-btn'),
    );
    component.getByDisplayValue((props.defaultQty + 1).toString());

    fireEvent.click(
      component.getByTestId('adjust-product-qty-box-increase-btn'),
    );
    component.getByDisplayValue((props.defaultQty + 1).toString());
  });
  test('it should match the style when render with mini type', () => {
    props.type = 'mini';
    const component = render(<AdjustProductQtyBox {...props} />);
    expect(component.getByTestId('adjust-product-qty-box')).toHaveStyle({
      height: '20px',
      lineHeight: '20px',
    });
    expect(
      component.getByTestId('adjust-product-qty-box-increase-btn'),
    ).toHaveStyle({
      width: '18.5px',
      height: '20px',
    });
  });
});
