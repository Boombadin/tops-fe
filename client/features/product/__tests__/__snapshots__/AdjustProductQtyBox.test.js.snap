// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features/product/components/AdjustProductQtyBox it should render default qty label from prop 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="AdjustProductQtyBox__IncrementInputBoxStyled-sc-15n8obw-0 cACVcL IncrementInputBox__Wrapper-sc-1o57gb-0 XRdDS"
        data-testid="adjust-product-qty-box"
        type="default"
      >
        <div
          class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
        >
          <div
            class="IncrementInputBox__IncButton-sc-1o57gb-4 bLReEN"
          >
            <button
              class="AdjustProductQtyBox__ButtonStyled-sc-15n8obw-1 cBQqxu"
              data-testid="adjust-product-qty-box-decrease-btn"
              type="default"
            >
              -
            </button>
          </div>
        </div>
        <div
          class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
        >
          <div
            class="IncrementInputBox__InputWrapper-sc-1o57gb-2 ieBGjh"
          >
            <input
              class="inputBox IncrementInputBox__InputBox-sc-1o57gb-3 bLAQDR"
              id="inputBox"
              readonly=""
              type="number"
              value="10"
            />
          </div>
        </div>
        <div
          class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
        >
          <div
            class="IncrementInputBox__IncButton-sc-1o57gb-4 bLReEN"
          >
            <button
              class="AdjustProductQtyBox__ButtonStyled-sc-15n8obw-1 cBQqxu"
              data-testid="adjust-product-qty-box-increase-btn"
              type="default"
            >
              +
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="AdjustProductQtyBox__IncrementInputBoxStyled-sc-15n8obw-0 cACVcL IncrementInputBox__Wrapper-sc-1o57gb-0 XRdDS"
      data-testid="adjust-product-qty-box"
      type="default"
    >
      <div
        class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
      >
        <div
          class="IncrementInputBox__IncButton-sc-1o57gb-4 bLReEN"
        >
          <button
            class="AdjustProductQtyBox__ButtonStyled-sc-15n8obw-1 cBQqxu"
            data-testid="adjust-product-qty-box-decrease-btn"
            type="default"
          >
            -
          </button>
        </div>
      </div>
      <div
        class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
      >
        <div
          class="IncrementInputBox__InputWrapper-sc-1o57gb-2 ieBGjh"
        >
          <input
            class="inputBox IncrementInputBox__InputBox-sc-1o57gb-3 bLAQDR"
            id="inputBox"
            readonly=""
            type="number"
            value="10"
          />
        </div>
      </div>
      <div
        class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
      >
        <div
          class="IncrementInputBox__IncButton-sc-1o57gb-4 bLReEN"
        >
          <button
            class="AdjustProductQtyBox__ButtonStyled-sc-15n8obw-1 cBQqxu"
            data-testid="adjust-product-qty-box-increase-btn"
            type="default"
          >
            +
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`features/product/components/AdjustProductQtyBox it should render readonly input 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="AdjustProductQtyBox__IncrementInputBoxStyled-sc-15n8obw-0 cACVcL IncrementInputBox__Wrapper-sc-1o57gb-0 XRdDS"
        data-testid="adjust-product-qty-box"
        type="default"
      >
        <div
          class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
        >
          <div
            class="IncrementInputBox__IncButton-sc-1o57gb-4 bLReEN"
          >
            <button
              class="AdjustProductQtyBox__ButtonStyled-sc-15n8obw-1 cBQqxu"
              data-testid="adjust-product-qty-box-decrease-btn"
              type="default"
            >
              -
            </button>
          </div>
        </div>
        <div
          class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
        >
          <div
            class="IncrementInputBox__InputWrapper-sc-1o57gb-2 ieBGjh"
          >
            <input
              class="inputBox IncrementInputBox__InputBox-sc-1o57gb-3 bLAQDR"
              id="inputBox"
              readonly=""
              type="number"
              value="1"
            />
          </div>
        </div>
        <div
          class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
        >
          <div
            class="IncrementInputBox__IncButton-sc-1o57gb-4 bLReEN"
          >
            <button
              class="AdjustProductQtyBox__ButtonStyled-sc-15n8obw-1 cBQqxu"
              data-testid="adjust-product-qty-box-increase-btn"
              type="default"
            >
              +
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="AdjustProductQtyBox__IncrementInputBoxStyled-sc-15n8obw-0 cACVcL IncrementInputBox__Wrapper-sc-1o57gb-0 XRdDS"
      data-testid="adjust-product-qty-box"
      type="default"
    >
      <div
        class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
      >
        <div
          class="IncrementInputBox__IncButton-sc-1o57gb-4 bLReEN"
        >
          <button
            class="AdjustProductQtyBox__ButtonStyled-sc-15n8obw-1 cBQqxu"
            data-testid="adjust-product-qty-box-decrease-btn"
            type="default"
          >
            -
          </button>
        </div>
      </div>
      <div
        class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
      >
        <div
          class="IncrementInputBox__InputWrapper-sc-1o57gb-2 ieBGjh"
        >
          <input
            class="inputBox IncrementInputBox__InputBox-sc-1o57gb-3 bLAQDR"
            id="inputBox"
            readonly=""
            type="number"
            value="1"
          />
        </div>
      </div>
      <div
        class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
      >
        <div
          class="IncrementInputBox__IncButton-sc-1o57gb-4 bLReEN"
        >
          <button
            class="AdjustProductQtyBox__ButtonStyled-sc-15n8obw-1 cBQqxu"
            data-testid="adjust-product-qty-box-increase-btn"
            type="default"
          >
            +
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
